Prompt user for credentials when http request needs it

arthmis opened the pull request:
<!-- Please describe your changes on the following line: -->
This adds two prompts that will ask the user for their username and password if a http request requires it.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #27439  (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

arthmis committed:
prompt user to get their credentials

Signed-off-by: Lloyd Massiah artmis9@protonmail.com

move credential prompt to a function

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented:
cc @jdm I asked a few questions about how to deal with errors in the comments.

arthmis commented on components/net/http_loader.rs:
```diff
@@ -120,6 +123,7 @@ impl Default for HttpState {
                 override_manager.clone(),
             )),
             override_manager,
+            embedder_proxy: Mutex::new(None),
```
I needed to make proxy since this `default` implementation is used in tests and I couldn't find an easy way to create an embedder for the default implementation. Is that fine?

jdm replied:
I think we should remove this Default implementation from the engine if it's only used in tests, and make a test helper that can provide a meaningful value. Probably we should store a trait object, so that tests can do something useful with it.

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1735,6 +1742,62 @@ impl Drop for ResponseEndTimer {
+fn prompt_user_for_credentials(
+    embedder_proxy: &Mutex<Option<EmbedderProxy>>,
+) -> Option<(Option<String>, Option<String>)> {
+    let Ok(guard) = embedder_proxy.lock() else {
```
should I just use `unwrap` on the `lock`?  I don't know if it's ok to panic here.

jdm replied:
Yeah, if locks are poisoned our network stack is going to fail imminently. Unwrapping is fine.

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1735,6 +1742,62 @@ impl Drop for ResponseEndTimer {
+        ));
+        ipc_receiver
+    };
+    let Ok(username) = username_receiver.recv() else {
```
when receiving the user input, if there is an error should we return the error or handle it here?

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1589,11 +1593,14 @@ async fn http_network_or_cache_fetch(
-            // We basically pretend that the user declined to enter credentials (#33616)
-            return response;
+            let credentials = prompt_user_for_credentials(&context.state.embedder_proxy);
+            if let Some((_username, _password)) = credentials {
```
if there are errors in this call for credentials how should we handle that? For now I chose to return None from the `prompt` function, but how should the `None` value be handled and is this even an appropriate way to deal with this?

arthmis committed:
add prompt for step 15.4

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1625,11 +1632,13 @@ async fn http_network_or_cache_fetch(
 
+        // TODO(#33616): Step 15.4 Prompt the end user as appropriate in request’s window
+        let credentials = prompt_user_for_credentials(&context.state.embedder_proxy);
+        if let Some((_username, _password)) = credentials {
```
Does it make sense to return `None` if there are errors when prompting the user? How should these errors be handled in general?

jdm replied:
Yes, I think logging the error and returning none is a reasonable strategy.

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1735,6 +1742,62 @@ impl Drop for ResponseEndTimer {
+        warn!("error getting password");
+        return None;
+    };
+
```
If I do add a new variant, would I have to modify the dialog box to show both `username` and `password` simultaneously?

arthmis replied:
oh ok, then maybe I will do that in a separate PR, since I think that would require some modification to the tinydialog library.

jdm replied:
I think the embedder side could still show two separate dialogs before returning data to the caller. I don't think we should spend time modifying tinydialogs unless it's absolutely necessary.

arthmis replied:
ah ok then I can do that.

arthmis committed:
add new prompt definition for user credentials

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
remove default implementation for HttpState which allowed making the embedder_proxy non-optional

- default implementation was only used in tests so created an alternative create_http_state function

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

add credentials to authentication cache

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented on components/net/tests/main.rs:
```diff
@@ -95,6 +95,25 @@ fn create_embedder_proxy() -> EmbedderProxy {
     }
 }
 
+fn create_http_state() -> HttpState {
```
I was able to remove the default implementation for `HttpState` and created this helper function that does the same thing but is only available in tests. Fortunately there was a `create_embedder_proxy` function which creates a dummy `embedder_proxy`. Thanks to this, there is no need for making `embedder_proxy` optional in `HttpState`

arthmis commented:
@jdm since this feature is about prompting users for authentication, are we assuming that the user will always be asked for a `username` and `password`? Is it possible to be asked for maybe single inputs like a `authentication key` or something like that? I looked [here](https://url.spec.whatwg.org/#include-credentials) and it seems including credentials always means `username` and `password`. Is that right?

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1735,6 +1718,36 @@ impl Drop for ResponseEndTimer {
+    let proxy = embedder_proxy.lock().unwrap();
+
+    let Ok((ipc_sender, ipc_receiver)) = ipc::channel() else {
+        error!("couldn't create ipc sender and receiver");
```
should I log the actual `Error` value?

jdm replied:
We can unwrap channel() errors. That's almost always going to be a resource starvation issue and the browser is going to crash anyways.

jdm commented:
> @jdm since this feature is about prompting users for authentication, are we assuming that the user will always be asked for a `username` and `password`? Is it possible to be asked for maybe single inputs like a `authentication key` or something like that? I looked [here](https://url.spec.whatwg.org/#include-credentials) and it seems including credentials always means `username` and `password`. Is that understanding correct?

Yes, we will only need to prompt for username and password.

arthmis commented:
@jdm do these particular changes need tests considering that they are only a part of the broader http request functionality?

And are there ways to provide localized strings during the prompts?

jdm commented:
> @jdm do these particular changes need tests considering that they are only a part of the broader http request functionality?
> 
> And are there ways to provide localized strings during the prompts?
> 
> Aside from that, I think this is ready for review

Yes, I believe this is not covered by the web-platform-tests because it involves user-agent interactions, so a unit test would be helpful here.

As for localized strings, we don't have a solution for that in servoshell at this time, so we can leave the hardcoded strings.

jdm reviewed (commented):
Looking good!

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1588,8 +1569,8 @@ async fn http_network_or_cache_fetch(
-            // TODO(#33616, #27439): Prompt the user for username and password from the window
-
+            let credentials = prompt_user_for_credentials(&context.state.embedder_proxy);
+            if let Some(_credentials) = credentials {}
```
not questioning whether I need to do this, but wondering where in the spec it says to do this?

jdm replied:
Oh, huh. The fetch spec says to.update the request url's username and password. However, from a user experience perspective, we would definitely want to make it easy for the user to avoid inputting their credentials on every request that is behind a basic auth gateway.

arthmis replied:
So I've implemented but it still infinite loops. And I assume you mean the next request is this line https://github.com/servo/servo/blob/11424f90b3cb0fe5688fce68139744f919c5be21/components/net/http_loader.rs#L1604

arthmis replied:
I'll push out my changes so you can see it better

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1588,12 +1569,21 @@ async fn http_network_or_cache_fetch(
+            let credentials = prompt_user_for_credentials(&context.state.embedder_proxy);
+            if let Some(credentials) = credentials {
+                // TODO(#33616): store the result as a proxy-authentication entry.
+                let entry = AuthCacheEntry {
```
how should we handle the username or password being `None`? From my testing if the user hits `cancel` while seeing this prompt then the returned value will be `None`.

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1588,12 +1569,21 @@ async fn http_network_or_cache_fetch(
+            if let Some(credentials) = credentials {
+                // TODO(#33616): store the result as a proxy-authentication entry.
+                let entry = AuthCacheEntry {
+                    user_name: credentials.username.unwrap(),
```
@jdm 

how should we handle the username or password being None? From my testing if the user hits cancel while seeing this prompt then the returned value will be None.

jdm replied:
If we don't have a username or password provided by the user, we can just keep the existing behaviour and return `response`.

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1588,12 +1569,21 @@ async fn http_network_or_cache_fetch(
+                };
 
+                let mut auth_cache = context.state.auth_cache.write().unwrap();
+                auth_cache.entries.insert("credentials".to_string(), entry);
```
so after inserting this, on line `1594` the `http_network_or_cache_fetch` gets called again. I assume thats what you meant by the new auth credentials entry will get picked up automatically. I'm testing this using the python example you put in #27439 so what should be the expected result?

jdm replied:
Since the auth cache is shared across all requests, the key we need to use is the [ascii serialization of the origin](https://github.com/servo/servo/blob/main/components/net/http_loader.rs#L438). You can get the appropriate origin from `request.current_url().origin()`.

arthmis replied:
ohh ok, thanks

arthmis replied:
ok that worked!!

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1588,12 +1569,31 @@ async fn http_network_or_cache_fetch(
-            // otherwise this creates an infinite loop
-            // We basically pretend that the user declined to enter credentials (#33616)
-            return response;
+            let (username, password) = {
+                let username = user_info_percent_encode(&username);
+                let password = credentials
+                    .password
+                    .map(|password| user_info_percent_encode(&password));
+
+                (username, password)
+            };
+
+            http_request
+                .current_url_mut()
+                .set_username(&username)
+                .unwrap();
+            http_request
+                .current_url_mut()
+                .set_password(password.as_deref())
+                .unwrap();
```
so I've updated this part to follow step 14.4 https://fetch.spec.whatwg.org/#ref-for-proxy-authentication-entry%E2%91%A0. Gotta scroll up a bit to see it. 

Let me know if I'm missing anything @jdm

arthmis commented on components/shared/net/lib.rs:
```diff
@@ -945,5 +945,31 @@ pub fn http_percent_encode(bytes: &[u8]) -> String {
     percent_encoding::percent_encode(bytes, HTTP_VALUE).to_string()
 }
 
+pub fn user_info_percent_encode<T: AsRef<[u8]>>(bytes: T) -> String {
```
I followed [the spec](https://url.spec.whatwg.org/#userinfo-percent-encode-set) but still a bit unsure if I did this right. 

Right now I'm not sure if the usage of `CONTROLS` conflicts with its usage in the function `http_percent_encode` found right about this function. I'm unsure because `CONTROLS` is a constant and I don't know if it retains characters added to it after its been called for future uses.

arthmis commented on components/shared/net/lib.rs:
```diff
@@ -945,5 +945,31 @@ pub fn http_percent_encode(bytes: &[u8]) -> String {
     percent_encoding::percent_encode(bytes, HTTP_VALUE).to_string()
 }
 
+pub fn user_info_percent_encode<T: AsRef<[u8]>>(bytes: T) -> String {
```
@jdm I followed [the spec](https://url.spec.whatwg.org/#userinfo-percent-encode-set) but still a bit unsure if I did this right. 

Right now I'm not sure if the usage of `CONTROLS` conflicts with its usage in the function `http_percent_encode` found right above this function. I'm unsure because `CONTROLS` is a constant and I don't know if it retains characters added to it after its been used and whether that affects future uses.

jdm replied:
I believe rust-url already implements this (which feels like the right layer of abstraction to me): https://github.com/servo/rust-url/blob/7492360d4230b67fa0e62794b6fde276525e5f84/url/src/lib.rs#L2301 . We should be able to call `set_username` on the url without performing any preprocessing on the username we pass in.

arthmis replied:
oh nice, perfect

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1588,12 +1569,28 @@ async fn http_network_or_cache_fetch(
+            http_request
+                .current_url_mut()
+                .set_username(&username)
+                .unwrap();
```
how should an error be handled here? Log and return response?

jdm replied:
I think that's reasonable.

arthmis replied:
I've added error handling, let me know if there is any other data that should be logged

arthmis commented on components/net/http_loader.rs:
```diff
@@ -1588,12 +1569,28 @@ async fn http_network_or_cache_fetch(
         // Step 14.3 If request’s use-URL-credentials flag is unset or isAuthenticationFetch is true, then:
         if !http_request.use_url_credentials || authentication_fetch_flag {
-            // TODO(#33616, #27439): Prompt the user for username and password from the window
+            let Some(credentials) = prompt_user_for_credentials(&context.state.embedder_proxy)
```
so I've updated this part to follow step 14.4 https://fetch.spec.whatwg.org/#ref-for-proxy-authentication-entry%E2%91%A0. Gotta scroll up a bit to see it. 

It's working so far. Let me know if I'm missing anything @jdm

jdm commented on components/net/tests/main.rs:
```diff
@@ -124,7 +124,7 @@ fn new_fetch_context(
 
     FetchContext {
-        state: Arc::new(create_http_state()),
+        state: Arc::new(create_http_state(Some(sender.clone()))),
```
Sorry for missing this previously, but since the embedder proxy is already part of FetchContext, why do we need to add it to HttpState? The code in http_loader.rs has access to a FetchContext.

arthmis replied:
you didn't miss it, I just added these changes yesterday. I'll look into it

arthmis replied:
I'm guessing we wouldn't need to add to the HttpState. I'll look into it. I was following the directions [laid out here](https://github.com/servo/servo/issues/27439#issue-667841183). So I wasn't aware that `FetchContext` already had access

jdm replied:
Well, I feel a bit silly now :D

arthmis replied:
So I didn't see anywhere that the `embedder_proxy` is part of the `FetchContext` besides within the `FileManager`. Unless you want to make the `embedder_proxy` part of the `FetchContext` instead of `HttpState`? I think that might make more sense conceptually. I haven't looked into how much effort it would be.

jdm replied:
You're right! I'm not sure how I misread the code.

arthmis committed:
add tests that are successful for the happy path

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
add test for user cancels prompt and user inputs incorrect credentials, and refactor shared code between tests

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
handle error when setting username and password in Url and ran formatting

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

renaming test functions

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented on components/net/tests/main.rs:
```diff
@@ -95,6 +97,76 @@ fn create_embedder_proxy() -> EmbedderProxy {
     }
 }
 
+fn create_embedder_proxy_and_receiver() -> (EmbedderProxy, EmbedderReceiver) {
```
this has some duplication with the function right above called `create_embedder_proxy`. I decided not to merge them but I'm open to doing so if you think it's better that way.

arthmis commented on components/net/tests/http_loader.rs:
```diff
@@ -1479,3 +1481,180 @@ fn test_origin_serialization_compatability() {
+}
+
+#[test]
+fn test_prompt_credentials_user_input_incorrect_credentials() {
```
should this test case cause an infinite loop? I was expecting an infinite loop because if the user makes the request with the wrong credentials, the user should be prompted again for credentials? Is that right?

jdm replied:
I believe that's because we pass `true` for the is_authentication_fetch argument when starting the next fetch, and we don't promo the user when that flag is true.

jdm replied:
But yes, I think multiple prompts are the desired outcome here.

arthmis commented:
After a successful review I'm going to squash a bunch of commits and then fix the DCO sign off issue

jdm reviewed (approved):
This is great work!

arthmis commented:
Not related to my changes, but there seems to be some panic when I run this test https://github.com/servo/servo/blob/6be0a64939c04fdcd9b8a11a863f1a04f07e4c9c/components/net/tests/http_loader.rs#L1344. And unfortunately since my tests are based off of this test, I have the same issue in my test cases. The tests do pass and there isn't any panic while using the mini browser. I ran this test on the `main` branch to confirm it's not an issue I introduced.

This is the error
```
thread 'tokio-runtime-1' panicked at /Users/art/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
A Tokio 1.x context was found, but it is being shutdown.
```

mrobinson commented:
> I'm getting an issue with `test-tidy`
> 
> ```
>   | /Users/art/Projects/servo/deny.toml:230: Unsoundness in `Iterator` and `DoubleEndedIterator` impls for `glib::VariantStrIter`: glib 0.19.9 registry+https://github.com/rust-lang/crates.io-index
> ```
> 
> though I didn't make any changes to anything related to this I think

This is #34764.

arthmis committed:
change authentication flag to false for proxy authentication. The spec doesn't specify that the flag should be true, and the flag is by default false

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
clean up test code a bit

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

jdm commented:
```
error[E0004]: non-exhaustive patterns: `PromptDefinition::Credentials(_)` not covered
   --> ports/servoshell/egl/servo_glue.rs:501:37
    |
501 |                     let res = match definition {
    |                                     ^^^^^^^^^^ pattern `PromptDefinition::Credentials(_)` not covered
    |
note: `PromptDefinition` defined here
   --> /home/runner/work/servo/servo/components/shared/embedder/lib.rs:125:1
    |
125 | pub enum PromptDefinition {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^
...
135 |     Credentials(IpcSender<PromptCredentialsInput>),
    |     ----------- not covered
    = note: the matched value is of type `PromptDefinition`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
    |
514 ~                         },
515 ~                         PromptDefinition::Credentials(_) => todo!(),
```

arthmis commented:
> ```
> error[E0004]: non-exhaustive patterns: `PromptDefinition::Credentials(_)` not covered
>    --> ports/servoshell/egl/servo_glue.rs:501:37
>     |
> 501 |                     let res = match definition {
>     |                                     ^^^^^^^^^^ pattern `PromptDefinition::Credentials(_)` not covered
>     |
> note: `PromptDefinition` defined here
>    --> /home/runner/work/servo/servo/components/shared/embedder/lib.rs:125:1
>     |
> 125 | pub enum PromptDefinition {
>     | ^^^^^^^^^^^^^^^^^^^^^^^^^
> ...
> 135 |     Credentials(IpcSender<PromptCredentialsInput>),
>     |     ----------- not covered
>     = note: the matched value is of type `PromptDefinition`
> help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
>     |
> 514 ~                         },
> 515 ~                         PromptDefinition::Credentials(_) => todo!(),
> ```

Oh wow, how did you end up running into that error?

arthmis commented:
@jdm So it seems to download the dev tools for Open Harmony OS I need to have a huawei account. So i can't build locally to test changes. I've implemented the skeleton for so the build would pass but I can't confirm it's fine until the CI tries it

arthmis committed:
add skeleton implementation to support open harmony and android

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented on ports/servoshell/egl/servo_glue.rs:
```diff
@@ -512,6 +513,9 @@ impl ServoGlue {
                             sender.send(cb.prompt_input(message, default, trusted))
                         },
+                        PromptDefinition::Credentials(_) => {
+                            warn!("implement credentials prompt for open harmony OS");
```
@jdm should I use the `todo!()` macro instead of a warning?

simonwuelker replied:
`todo!()` would cause servo to crash on ohos/android when a credential prompt is requested, that is definitely undesirable. A warning is perfectly fine I think.

arthmis replied:
yeah that makes sense, regarding `todo!()`. Would it be worth marking `PromptDefinition` non-exhaustive now?

simonwuelker replied:
> Would it be worth marking PromptDefinition non-exhaustive now?

I think it would be better to do that in a separate PR as it's unrelated to this one.

jschwe replied:
Note: This code is shared and used by both OpenHarmony and Android, so probably the warning message should reflect that. 
Both currently don't support prompts, so just issuing a warning is perfectly fine for this PR.

arthmis committed:
update warning message to include Android

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
fix build error for OH os and Android

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
remove unused import to fix warning

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented:
ok so second time this failed, I'll try and get the android environment running so I can actually run a build for android and that should allow me to get proper checking. I did add a `Ok(())` in the `PromptCredentials` branch which should fix the issue, and I fixed the unused import warning.

simonwuelker commented:
> ok so second time this failed, I'll try and get the android environment running so I can actually run a build for android and that should allow me to get proper checking. I did add a `Ok(())` in the `PromptCredentials` branch which should fix the issue, and I fixed the unused import warning.

Alternatively, you can run the android/ohos CI on your fork. (Not ideal, but easier than setting up android if you're not going to work on it in the future).
Under the "*Actions*" tab, choose the OpenHarmony workflow and start it manually.

arthmis commented:
> Alternatively, you can run the android/ohos CI on your fork. (Not ideal, but easier than setting up android if you're not going to work on it in the future). Under the "_Actions_" tab, choose the OpenHarmony workflow and start it manually.

Oh nice, that will save me a lot of trouble. Thanks

wusyong commented on components/net/http_loader.rs:
```diff
@@ -1737,6 +1760,29 @@ impl Drop for ResponseEndTimer {
+
+    let (ipc_sender, ipc_receiver) = ipc::channel().unwrap();
+
+    proxy.send((
+        None,
+        EmbedderMsg::Prompt(
+            PromptDefinition::Credentials(ipc_sender),
+            PromptOrigin::Trusted,
+        ),
+    ));
```
This should contain the ID of the browsing context. Otherwise, the embedder has no idea where it comes from.

jdm replied:
Could you open an issue?
